### PR TITLE
[WIP] RAG end-to-end retriever training (with ray workers)

### PR DESCRIPTION
**Pull request by shamanez**
# What does this PR do?

 As mentioned in this [issue](https://github.com/huggingface/transformers/issues/9646), this PR adds the fine-tuning ability the Retriever in the original RAG implementation. 

This PR first updates the ctx_encoder and then initializes the index using RAY workers.

@lhoestq 
@amogkam

### CODE REVIEW

File: src/transformers/models/rag/modeling_rag.py [hunk 1]
```diff
@@ -483,6 +484,7 @@ def __init__(
         question_encoder: Optional[PreTrainedModel] = None,
         generator: Optional[PreTrainedModel] = None,
         retriever: Optional = None,  # or maybe just use a `set_retriever(...)` method
+        ctx_encoder: Optional = None, #newly_added
```
Comment (lhoestq) [inline src/transformers/models/rag/modeling_rag.py:487]: Thanks for extending RAG to support backpropagation to the ctx_encoder :)

I think it's ok to allow users to optionally provide the context encoder if they want to backpropagate the gradient to the context encoder, but I'd like your opinion on this @patrickvonplaten 

In this case maybe we can have a parameter somewhere to toggle the computation of the ctx_encoder embeddings using the ctx_encoder instead of taking the embeddings from the retriever that don't backprogagate to the ctx_encoder. What do you think ?

Comment (shamanez) [reply src/transformers/models/rag/modeling_rag.py:487]: @lhoestq  Thanks a lot for getting into this :). I was thinking the same.  This can be done by adding an if condition in the doc score calculation part. If there's no ctx_encoder calculate the doc scores with retrieved embedding.

File: src/transformers/models/rag/modeling_rag.py [hunk 2]
```diff
@@ -595,10 +607,26 @@ def forward(
                 context_input_ids = context_input_ids.to(input_ids)
                 context_attention_mask = context_attention_mask.to(input_ids)
 
-                # compute doc_scores
-                doc_scores = torch.bmm(
+                retrived_doc_input_ids=retrived_doc_input_ids.to(input_ids)
+                retrived_doc_attention_mask=retrived_doc_attention_mask.to(input_ids)
+
+                grad_doc_embeddings=self.ctx_encoder(retrived_doc_input_ids,attention_mask=retrived_doc_attention_mask,
```
Comment (shamanez) [inline src/transformers/models/rag/modeling_rag.py:613]: @lhoestq  Here, we can  use and if the similar to 

```
if self.ctx_encoder is not none:
        grad_doc_embeddings=self.ctx_encoder(retrived_doc_input_ids,attention_mask=retrived_doc_attention_mask,

```

File: src/transformers/models/rag/modeling_rag.py [hunk 3]
```diff
@@ -550,8 +553,10 @@ def forward(
             >>> # initialize with RagRetriever to do everything in one forward call
             >>> model = RagModel.from_pretrained("facebook/rag-token-base", retriever=retriever)
 
-            >>> inputs = tokenizer("How many people live in Paris?", return_tensors="pt")
-            >>> outputs = model(input_ids=inputs["input_ids"])
+            >>> input_dict = tokenizer.prepare_seq2seq_batch("How many people live in Paris?", "In Paris, there are 10 million people.", return_tensors="pt")
```
Comment (patrickvonplaten) [inline src/transformers/models/rag/modeling_rag.py:556]: We have deprecated the `prepare_seq2seq_batch` function. Could you maybe take a look here: https://github.com/huggingface/transformers/blob/0d909f6bd8ca0bc1ec8f42e089b64b4fffc4d230/src/transformers/models/rag/modeling_rag.py#L1208 at how to handle it now?

File: src/transformers/models/rag/modeling_rag.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/rag/modeling_rag.py:487]: I think instead of optionally providing the `ctx_encoder` to `init` it would be nicer to make a setter method instead, something like:

```python
def set_context_encoder_for_training(...)
```

This is cleaner IMO. For composite models, I'd really like to limit the __init__ only to necessary components (being the `question_encoder`, `generator`, and `config`) and don't want to add any optional ones that are only used for training. It makes the model less understandable IMO.

What do you guys think?

Comment (shamanez) [reply src/transformers/models/rag/modeling_rag.py:487]: me too. Will do it.

**Comment by shamanez** [comment]
@lhoestq  @patrickvonplaten 

I have already started doing the above changes you have mentioned.  Apart from that, I changed the following elements of the codebase. 

1. I used a dedicated RAY worker to compute embeddings for the dataset with an updated ctx encoder in this version.  However, the process of add_faiss_index gets very slow when running inside a ray worker (I think it is due to the need for multiprocessing threads). I tried to increase the number of CPU cores, but it is still very slow.  The computing of embeddings is an embarrassingly parallel task, where we can share the dataset between GPUs and compute them very fast. Nevertheless, it is hard to work with RAY when it comes to multiple GPUs. So I utilize the DDP process to compute embeddings using N number of dedicated GPUs that only do the embeddings calculation task.

2. Then I did a minor thing. Pytorch lightning has removed the DDP accelerators in their latest installation. Nevertheless, we can easily use the **on_sanity_check_start** callback to initialize the index when using RAY. I feel it is a lot cleaner.  

__________________________________________________________________________________________________

As per my experiments, at the moment end-to-end training process is stable.  I would love to double-check the following parts with your help. 


1. Re-loading of the updated index for the workers.
2. Re-initialization of the retrieval index.

Apart from them, I see **add_faiss_index** can get hours when the dataset consists of more than a million passages. My custom dataset has 8 million datasets. Is it normal, or should we able to improve it?   If we can improve its speed, this whole process can be very engineering friendly. 

Please let me know your ideas. I will quickly do the updated PR.

**Comment by lhoestq** [comment]
For the records, we are discussing the indexing speed difference here: https://github.com/huggingface/datasets/issues/2046

**Comment by shamanez** [comment]
@lhoestq 

I and @elliott-wen updated the codebase. Now the embedding update happens with a parallel process and we use the stale gradients to update the entire model (pretty-much similar to REALM).

**Comment by patrickvonplaten** [comment]
@shamanez do you need any help to get the test working?

**Comment by shamanez** [comment]
Hi @patrickvonplaten  I did change the code and got a stable end to end trainable RAG. I also did the changes you have mentioned. But I had to update the code with new pytorch lightning , specially they do not use plugging now.  So is that ok to upload the code with  new end-to-end fine-tune script and lightning base ?

**Comment by lhoestq** [comment]
Hi, that's good news !

> is that ok to upload the code with a new end-to-end fine-tune script and lightning base?

I think it could be a good idea to make the code compatible with the latest pytorch-lightning yes :)
Especially since many things we used in lightning_base don't work anymore, and that we can now hope pytorch-lightning to not do such radical changes again.
pinging @patrickvonplaten to confirm it's ok

**Comment by shamanez** [comment]
@lhoestq  Thanks. 

BTW I read this new paper named [Retrieval Augmentation Reduces Hallucination in Conversation](https://arxiv.org/abs/2104.07567), which kinds of highlights the importance of RAG-like models in LM modeling.  So I do believe end-to-end fine-tuning can allow users to experiment with different components of the RAG architecture.


Since you guys helped me a lot in this process, is it okay to include your names in the blog? Here's a link to the unpublished draft blog post.



https://medium.com/@shamanesiriwardhana/end-to-end-rag-fine-tuning-with-huggingface-pytorch-lightning-and-ray-4b4385322552



Please let me know your thoughts.

**Comment by shamanez** [comment]
Thanks.

**Comment by shamanez** [comment]
> Hi, that's good news !
> 
> > is that ok to upload the code with a new end-to-end fine-tune script and lightning base?
> 
> I think it could be a good idea to make the code compatible with the latest pytorch-lightning yes :)
> Especially since many things we used in lightning_base don't work anymore, and that we can now hope pytorch-lightning to not do such radical changes again.
> pinging @patrickvonplaten to confirm it's ok

So is that ok we create a folder in researc_project naming, RAG-end-to-end-Retriever training

**Comment by shamanez** [comment]
closing this with a new pull request. 

https://github.com/huggingface/transformers/pull/11655

**Comment by patrickvonplaten** [comment]
Hey @shamanez,

Sorry for being so inactive here! I reviewed your newly opened PR :-)

**Comment by shamanez** [comment]
Hey, it is totally fine. Thanks million times :).
